content/advent-2015/reducing-boilerplate-with-go-generate.md

```diff
@@ -0,0 +1,298 @@
++++
+author = ["Ernesto Jiménez"]
+date = "2015-12-10T08:00:00+00:00"
+title = "Reducing boilerplate with go generate"
+series = ["Advent 2015"]
++++
+
+Go is an awesome language. It's simple, powerful, has great tooling and
+many of us really enjoy using it every day. However, as it usually
+happens with strongly typed languages, we write a good deal of
+boilerplate to connect things around.
+
+In this post we'll cover mostly three points:
+
+1. Why can we build tools with Go that will help reduce boilerplate using code generation.
+2. What are the building blocks for code generation in Go.
+3. Where can we find examples of code generation tools to learn more.
+
+# Why use code generation to reduce boilerplate?
+
+Sometimes we try to reduce boilerplate by using reflection and filling
+our projects with methods accepting `interface{}`. However, whenever a
+method takes an `interface{}`, we are throwing our type safety out of
+the window. When using type assertions and reflection, the compiler is
+unable to check we are passing the right types and we are more exposed
+to runtime panics.
+
+Some of the boilerplate code we've got can be mostly inferred from the
+code we already have in our project. For that, we can write tools that
+will read our project's source code and generate the relevant code.
+
+# The building blocks to code generation
+
+## Reading code
+
+The standard library has a wonderful set of packages ready to do most of
+the heavy lifting when it comes to reading and parsing code.
+
+* `go/build`: gathers information about a go package. Given a package
```

Comment:
what do you think about linking these to the documentation?

eg. https://golang.org/pkg/go/build/